File: ios/App/App/FolioleCompanionSyncPackPayloadWriter.swift
import Foundation

enum FolioleCompanionSyncPackPayloadWriter {
    static func copy(_ database: FolioleCompanionSyncPackSQLite, plans: [[String: Any]]) throws {
        var payloads: [String: String] = [:]
        for plan in plans {
            guard let objectType = plan["objectType"] as? String, let sql = plan["sql"] as? String else { continue }
            for row in try database.namedRows(sql) {
                guard let objectId = row["__object_id"] as? String else { continue }
                var payload: [String: Any] = [:]
                for (name, value) in row where name != "__object_id" { assign(&payload, name, value) }
                let data = try JSONSerialization.data(withJSONObject: payload, options: [.sortedKeys])
                payloads[key(objectType, objectId)] = String(decoding: data, as: UTF8.self)
            }
        }
        let states = try database.rows(
            "SELECT object_type, object_id, content_hash, updated_at, deleted_at FROM sync_object_state " +
            "WHERE object_type NOT IN ('external_document','node')"
        )
        for state in states {
            guard state.count == 5, let type = state[0] as? String, let objectId = state[1] as? String else { continue }
            let deleted = state[4] as? String
            let payload = deleted == nil ? payloads[key(type, objectId)] : nil
            if deleted == nil && payload == nil { continue }
            try database.insertSyncObject([type, objectId, state[2], payload, state[3], deleted])
        }
    }

    private static func assign(_ result: inout [String: Any], _ name: String, _ value: Any) {
        guard let separator = name.range(of: "__") else { result[name] = value; return }
        let parent = String(name[..<separator.lowerBound]), child = String(name[separator.upperBound...])
        var nested = result[parent] as? [String: Any] ?? [:]
        nested[child] = value
        result[parent] = nested
    }

    private static func key(_ type: String, _ id: String) -> String { type + "\u{0}" + id }
}

File: android/app/src/main/java/com/foliole/android/FolioleCompanionSyncPackPayloadWriter.java
package com.foliole.android;

import android.content.ContentValues;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.HashMap;
import java.util.Map;

final class FolioleCompanionSyncPackPayloadWriter {
    private FolioleCompanionSyncPackPayloadWriter() {}

    static void copy(SQLiteDatabase pack, JSONArray plans) throws Exception {
        Map<String, String> payloads = loadPayloads(pack, plans);
        try (Cursor states = pack.rawQuery(
            "SELECT object_type, object_id, content_hash, updated_at, deleted_at FROM sync_object_state " +
                "WHERE object_type NOT IN ('external_document','node')", null
        )) {
            while (states.moveToNext()) copyState(pack, payloads, states);
        }
    }

    private static Map<String, String> loadPayloads(SQLiteDatabase pack, JSONArray plans) throws Exception {
        Map<String, String> payloads = new HashMap<>();
        for (int index = 0; index < plans.length(); index++) {
            JSONObject plan = plans.getJSONObject(index);
            String objectType = plan.getString("objectType");
            try (Cursor rows = pack.rawQuery(plan.getString("sql"), null)) {
                int objectIdIndex = rows.getColumnIndexOrThrow("__object_id");
                while (rows.moveToNext()) {
                    payloads.put(key(objectType, rows.getString(objectIdIndex)), json(rows).toString());
                }
            }
        }
        return payloads;
    }

    private static void copyState(SQLiteDatabase pack, Map<String, String> payloads, Cursor state) throws Exception {
        String objectType = state.getString(0);
        String objectId = state.getString(1);
        String payload = state.isNull(4) ? payloads.get(key(objectType, objectId)) : null;
        if (state.isNull(4) && payload == null) return;
        ContentValues values = new ContentValues();
        values.put("object_type", objectType); values.put("object_id", objectId);
        values.put("content_hash", state.getString(2)); values.put("updated_at", state.getString(3));
        if (payload == null) values.putNull("payload_json"); else values.put("payload_json", payload);
        if (state.isNull(4)) values.putNull("deleted_at"); else values.put("deleted_at", state.getString(4));
        pack.insertOrThrow("sync_objects", null, values);
    }

    private static JSONObject json(Cursor row) throws Exception {
        JSONObject result = new JSONObject();
        for (int index = 0; index < row.getColumnCount(); index++) {
            if ("__object_id".equals(row.getColumnName(index))) continue;
            String[] path = row.getColumnName(index).split("__", 2);
            JSONObject target = result;
            if (path.length == 2) {
                if (!result.has(path[0])) result.put(path[0], new JSONObject());
                target = result.getJSONObject(path[0]);
            }
            target.put(path[path.length - 1], value(row, index));
        }
        return result;
    }

    private static String key(String objectType, String objectId) {
        return objectType + "\u0000" + objectId;
    }

    private static Object value(Cursor row, int index) {
        switch (row.getType(index)) {
            case Cursor.FIELD_TYPE_NULL: return JSONObject.NULL;
            case Cursor.FIELD_TYPE_INTEGER: return row.getLong(index);
            case Cursor.FIELD_TYPE_FLOAT: return row.getDouble(index);
            default: return row.getString(index);
        }
    }
}
